## Introduction

ALLHiC_adjuster is a toolset contain serval scripts for adjusting ALLHiC result.

## Dependencies

### Software

[jcvi](<https://github.com/tanghaibao/jcvi>)

[ALLHiC](<https://github.com/tangerzhang/ALLHiC>)

### Modules

matplotlib

## Usage

### Step 1. Get linkage from anchors file generated by jcvi

```bash
ConvertAnchors.py X.bed Y.bed X.Y.anchors X.Y.link
```

### Step 2. Get blocks with link

```bash
GetBreakBlock.py X.Y.link 20 X.Y.blocks
```

20 means 1/20 of chromosome's length, this is a threshold value that used for clustering contigs into different blocks.

### Step 3. Draw dotplot

```bash
DrawDotPlotColor.py X.Y.link X.Y.blocks groups.agp 20 X.Y.blocks.pdf
```

20 is the threshold that same as step 2.

### Step 4. Adjust

After step 3, you will get a pdf showed blocks with red line, and with the first and last contig id at both ends of line and a file named *ctg_blocks.txt* with all contigs in different blocks. 

Then, use scripts below to adjust tour files.

```bash
# split tour with contig id
split_tour.py groupX.tour tig_X1,tigX2
# reverse tour
reverse_tour.py groupX.tour
# merge tours
merge_tour.py groupX_1.tour,groupX_2.tour groupX.tour
```

### Step 5. Build

Make sure the folder for building only contain necessary .tour files then run:

```bash
ALLHiC_build draft.asm.fasta
```

draft.asm.fasta is contig-level assmbly

## Otherwise

**ALLHiC_mono_allele_minimap.py** is a script for generating allele.table.txt with a haplotype

**convert_agp_to_tour.py** is a script for converting groups.agp to several tours

**ConvertAnchorsForCircos.py** is a script for converting anchors to linkages that can be used in circos

**ConvertRagooOrderingtoAgp.py** is a script that can convert ragoo orderings to agp file

**ConvertRagooOrderingtoTour.py** is a script that can convert a single ragoo ordering file to tour file

